### ISSUE TITLE
AdamW Optimizer formula in Section 4.3 differs from original

### ISSUE THREAD

**Issue by artbataev** [stanford-cs336/assignment1-basics#54]
I Section 4.3 in the pseudocode applying weight decay in the last statement uses already updated parameters (from the previous line). 
<img width="550" height="260" alt="Image" src="https://github.com/user-attachments/assets/3ce5cc31-48e4-4265-9a58-dec76d3737d0" />

It looks like the original paper https://arxiv.org/abs/1711.05101 and PyTorch https://docs.pytorch.org/docs/stable/generated/torch.optim.AdamW.html implementation use **previous** parameters to apply weight decay.

So, instead of 
$\theta \leftarrow \theta - \alpha_{t} \frac{m}{\sqrt{v} + \epsilon}$
$\theta \leftarrow \theta - \alpha \lambda \theta$ ($\theta$ is updated previously)

there should be
$\theta \leftarrow \theta - \alpha \lambda \theta$
$\theta \leftarrow \theta - \alpha_{t} \frac{m}{\sqrt{v} + \epsilon}$
or (in one statement)
$\theta \leftarrow \theta - \alpha_{t} \frac{m}{\sqrt{v} + \epsilon} - \alpha \lambda \theta$

The tests are passed with both implementations.
Is this an error, or was the change done intentionally?

**Comment by capyun007** [comment]
nice
